providers/ms_entra_id.go

```diff
@@ -89,6 +95,65 @@ func (p *MicrosoftEntraIDProvider) ValidateSession(ctx context.Context, session
 	return p.OIDCProvider.ValidateSession(ctx, session)
 }
 
+// Redeem exchanges the OAuth2 authentication token for an ID token, considering federated token authentication
+func (p *MicrosoftEntraIDProvider) Redeem(ctx context.Context, redirectURL, code, codeVerifier string) (*sessions.SessionState, error) {
+
+	// Perform usual Redeem for client secret authentication
+	if !p.federatedTokenAuthEnabled {
+		return p.OIDCProvider.Redeem(ctx, redirectURL, code, codeVerifier)
+	}
+
+	return p.redeemWithFederatedToken(ctx, redirectURL, code, codeVerifier)
+}
+
+// redeemWithFederatedToken performs custom token exchange with federated token instead of client secret
+func (p *MicrosoftEntraIDProvider) redeemWithFederatedToken(ctx context.Context, redirectURL, code, codeVerifier string) (*sessions.SessionState, error) {
+	params := url.Values{}
+
+	// parse federated token
+	federatedTokenPath := os.Getenv("AZURE_FEDERATED_TOKEN_FILE")
+	federatedToken, err := os.ReadFile(federatedTokenPath)
+	if err != nil {
+		return nil, fmt.Errorf("error reading federated token file %s: %s", federatedTokenPath, err)
+	}
+
+	// create custom exchange parameters
+	if codeVerifier != "" {
+		params.Add("code_verifier", codeVerifier)
+	}
+	params.Add("redirect_uri", redirectURL)
+	params.Add("client_id", p.ClientID)
+	params.Add("client_assertion", string(federatedToken))
+	params.Add("client_assertion_type", "urn:ietf:params:oauth:client-assertion-type:jwt-bearer")
+	params.Add("code", code)
+	params.Add("grant_type", "authorization_code")
+
+	// perform exchange
+	resp := requests.New(p.RedeemURL.String()).
+		WithContext(ctx).
+		WithMethod("POST").
+		WithBody(bytes.NewBufferString(params.Encode())).
+		SetHeader("Content-Type", "application/x-www-form-urlencoded").
+		Do()
+
+	// prepare token of type *oauth2.Token
+	body := resp.Body()
+
+	var token *oauth2.Token
+	var rawResponse interface{}
+
+	if err := json.Unmarshal(body, &rawResponse); err != nil {
+		return nil, err
+	}
+
+	if err := json.Unmarshal(body, &token); err != nil {
+		return nil, err
+	}
+
+	// create session using new token and generic OIDC provider
+	return p.OIDCProvider.createSession(ctx, token.WithExtra(rawResponse), false)
```

Comment:
Why is the rawResponse necessary to have as part of the session creation?

Reply:
Without rawResponse, the idToken isn't included in the Token object as it's not a part of the default oauth2 type and session creation is failing. This is reproducing this behavior:
https://github.com/golang/oauth2/blob/master/internal/token.go#L317

I've confirmed it by dumping content of token object returned by default OIDC provider - raw response is included.